### PR TITLE
Line heuristic

### PR DESCRIPTION
**Pull request by ptersilie**
This PR adds a line-based heuristic for automatic language box detection. In order to find locations at which language boxes can be inserted, this heuristic simply scans all tokens up to the beginning of the line. While this heuristic performs really well given some preliminary experiments, there are still edge cases where the other heuristics are needed, for example in Java+PHP:
```
class x{
    function x(){
        $x = 1; // Error happens in this line
    }
}
```

This PR also moves the configuration of which heuristic is to be used to a new config module `config.py` to which we will move over all other editor configuration options in a separate PR.

Update: This PR now also runs all heuristics and then ranks their results by parsing distance, discarding non-maximal results.

### CODE REVIEW

**Comment by ptersilie** [comment]
This PR is currently on hold, as the line-based approach doesn't work in the following composition of Java and PHP:
```Java
class X {
    int x = !e($x) ? $y : $z == 2; // this used to be 1 == 2
}
```
The line-based approach first creates a box around `!e($x)` and then around `$y == 2`. But when we continue typing `: $z`, the first box doesn't get extended. For this to work we need to first add another tweak that allows the automatic lbox detector to break down the second language box when trying to extend the first one.

**Comment by ptersilie** [comment]
I have updated this PR and it is now ready for review.

**Comment by ltratt** [comment]
Are the comments at the top of the PR still correct / relevant?

**Comment by ptersilie** [comment]
To keep the PR small I have decided to ignore the broken test this causes and do the fix in a separate PR.

**Comment by ptersilie** [comment]
> Are the comments at the top of the PR still correct / relevant?

I have added an update to the PR description.

File: lib/eco/autolboxdetector.py
Comment (ltratt) [inline lib/eco/autolboxdetector.py]: This formatting is very hard to read. Maybe too much indentation or ...?

File: lib/eco/autolboxdetector.py
Comment (ltratt) [inline lib/eco/autolboxdetector.py]: An obvious question is what happens if we remove the `not valid and` here (and remove duplicates)?

Comment (ltratt) [reply lib/eco/autolboxdetector.py]: Ah, OK, this is addressed in the following commit.

File: lib/eco/autolboxdetector.py
Comment (ltratt) [inline lib/eco/autolboxdetector.py]: `for start, _, _, dist in valid`?

File: lib/eco/treemanager.py
Comment (ltratt) [inline lib/eco/treemanager.py]: `XXX`?

File: lib/eco/autolboxdetector.py
Comment (ptersilie) [reply lib/eco/autolboxdetector.py]: I'm not sure how I can improve this.

Comment (ltratt) [reply lib/eco/autolboxdetector.py]: A nested `if` would probably be less worse. I had to read this 3 or 4 times to work out what was going on.

File: lib/eco/treemanager.py
Comment (ptersilie) [reply lib/eco/treemanager.py]: This was a reminder that I need to filter the possible results for expanding a language box by removing all non-maximal results (similar to the creating of language boxes). However, now that I have implemented this I am unsure if that is what we want. Because it means that we will always just expand to the longest match and not show other possible expansion options. Thoughts?

Comment (ltratt) [reply lib/eco/treemanager.py]: We might be able to determine this experimentally: if you run the experiment with both, does the number of acceptable outcomes differ in a meaningful way?

Comment (ptersilie) [reply lib/eco/treemanager.py]: I can try. We will likely see less "No insertion (Multi)", since we will always pick the longest match. This won't tell us though if the other options are something the user wanted to do since they may not always want to expand to the longest match.

Comment (ltratt) [reply lib/eco/treemanager.py]: Your call. I don't really know what's best, either.

Comment (ptersilie) [reply lib/eco/treemanager.py]: I'm always a bit reluctant removing options from the user, so I think I'm just gonna leave it as is for now and remove this comment. I've saved the filtering code, so we can bring it back if we decide it's the better option.

Comment (ltratt) [reply lib/eco/treemanager.py]: I think we should run the experiment, because it's an almost zero-cost thing to do. If, as is likely, it doesn't tell us very much then we can keep things as-is. Otherwise, AFAICS, we'll never have any other data to guide our decision...

**Comment by ptersilie** [comment]
See if these make things better!

File: lib/eco/treemanager.py
Comment (ptersilie) [reply lib/eco/treemanager.py]: Okay, I'll start a run and see what happens.

File: lib/eco/autolboxdetector.py
Comment (ltratt) [inline lib/eco/autolboxdetector.py:81]: ISTR there was another one of these a few lines below that's probably worth the same treatment.

Comment (ptersilie) [reply lib/eco/autolboxdetector.py:81]: I think that one uses all elements of the tuple.

Comment (ltratt) [reply lib/eco/autolboxdetector.py:81]: Ah OK.

**Comment by ltratt** [comment]
If the test pass, please squash.

File: lib/eco/treemanager.py
Comment (ptersilie) [reply lib/eco/treemanager.py]: Looking at the results, this change didn't really seem to improve our overall results, so I'm just leaving things as is, as I believe multiple options are in this case better than no options.

**Comment by ptersilie** [comment]
Squashed!

**Review by ltratt — APPROVED** [review]

**Pull request merged**
